pyop2/codegen/rep2loopy.py

```diff
@@ -205,16 +208,18 @@ def __call__(self, preamble_info):
         yield ("0", self.preamble)
 
 
+@dataclass(frozen=True, init=False)
 class PyOP2KernelCallable(loopy.ScalarCallable):
     """Handles PyOP2 Kernel passed in as a string
     """
 
-    fields = set(["name", "parameters", "arg_id_to_dtype", "arg_id_to_descr", "name_in_target"])
     init_arg_names = ("name", "parameters", "arg_id_to_dtype", "arg_id_to_descr", "name_in_target")
 
+    parameters: tuple
+
     def __init__(self, name, parameters, arg_id_to_dtype=None, arg_id_to_descr=None, name_in_target=None):
-        super(PyOP2KernelCallable, self).__init__(name, arg_id_to_dtype, arg_id_to_descr, name_in_target)
-        self.parameters = parameters
+        super().__init__(name, arg_id_to_dtype, arg_id_to_descr, name_in_target)
+        object.__setattr__(self, "parameters", tuple(parameters))
```

Comment:
Why `object.__setattr__` and not `self.parameters = ...` or `setattr`?

Reply:
Because this is a frozen dataclass. I was copying what loopy do internally ([example](https://github.com/inducer/loopy/blob/76f4afc27f343bc7e95a11b9b04d30c574402bcd/loopy/symbolic.py#L646), I really don't like it).

Reply:
Does it break otherwise? Is there something specific about frozen dataclasses that means you have to do this?
I would usually advocate for using the smallest hammer possible!

Reply:
Yeah otherwise it breaks unfortunately

Reply:
`setattr` breaks or `self.parameters = `? What's the error?

Reply:
My guess is that dataclasses overload `__setattr__` and so we need to specify `object.__setattr__`. Will report back.

Reply:
That's exactly right. There's no *real* immutability in Python, and so the MO of frozen dataclasses is to make mutation stick out like a sore thumb. Reading your comments above, I would say mission accomplished. :slightly_smiling_face: 

The conventional, easier-on-the-eyes approach would be to just delete the constructor (and let the `@dataclass` transform worry about writing they eyesore code). This is not feasible here because the argument order would mismatch what dataclasses generates, breaking compatibility. :cry:

Reply:
I see. Not the nicest thing but doesn't sound like there's much alternative if this is a dataclass but the `PyOP2KernelCallable` has a different signature to `PyOP2KernelCallable`.